crates/ra_hir/src/ty.rs

```diff
@@ -906,7 +906,16 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
                 }
             }
             ast::Expr::RangeExpr(_e) => Ty::Unknown,
-            ast::Expr::BinExpr(_e) => Ty::Unknown,
+            ast::Expr::BinExpr(e) => match e.op() {
```

Comment:
You should recursively call `infer_expr` on `lhs` and `rhs` here, otherwise we won't infer any types for the subexpressions. For `&&` and `||`, we can even pass down the expectation that the operands should be bools (i.e. `Expectation::has_type(Ty::Bool)`. (Though that doesn't really make a difference yet, as long as we don't have coercion.)